File: Reference/DontStarveScript/stategraph.lua
require("class")

StateGraphWrangler = Class(function(self)
        self.instances = {}
        self.updaters = {}
        self.tickwaiters = {}
        self.hibernaters = {}
        self.idlers = {}
        self.haveEvents = {}
end)

SGManager = StateGraphWrangler()

function StateGraphWrangler:SendToList(inst, list, allowunhibernate)
    local old_list = self.instances[inst]
    if old_list then
        if not allowunhibernate and old_list == self.hibernaters then
            return
        end
        old_list[inst] = nil
    end
    
    self.instances[inst] = list
    
    if list then
        list[inst] = true
    end
end

function SGManager:OnEnterNewState(inst)
    if self.instances[inst] then
        self:SendToList(inst, self.updaters)
    end
end

function StateGraphWrangler:OnSetTimeout(inst)
    if self.instances[inst] then
        self:SendToList(inst, self.updaters)
    end
end

function StateGraphWrangler:OnPushEvent(inst)
    if self.instances[inst] then
        self.haveEvents[inst] = true
        return true
    end
    return false
end

function StateGraphWrangler:Hibernate(inst)
    if self.instances[inst] then
        self:SendToList(inst, self.hibernaters)
    end
end

function StateGraphWrangler:Idle(inst)
    if self.instances[inst] then
        self:SendToList(inst, self.idlers)
    end
end

function StateGraphWrangler:Wake(inst)
    if self.instances[inst] then
       self:SendToList(inst, self.updaters, true)
    end
end

function StateGraphWrangler:Sleep(inst, time_to_wait)
    if self.instances[inst] then
        local sleep_ticks = time_to_wait/GetTickTime()
        if sleep_ticks == 0 then sleep_ticks = 1 end

        local target_tick = math.floor(GetTick() + sleep_ticks) + 1
        local waiters = self.tickwaiters[target_tick]

        if not waiters then
            waiters = {}
            self.tickwaiters[target_tick] = waiters
        end
        self:SendToList(inst, waiters)
    end
end

function StateGraphWrangler:OnRemoveEntity(inst)
    if self.instances[inst.sg] then
        self:RemoveInstance(inst.sg)
    end
end

function StateGraphWrangler:RemoveInstance(inst)
    local old_list = self.instances[inst]
    if old_list ~= nil then
        old_list[inst] = nil
    end

    self.instances[inst] = nil
    self.updaters[inst] = nil
    self.tickwaiters[inst] = nil
    self.hibernaters[inst] = nil
    self.haveEvents[inst] = nil
end

function StateGraphWrangler:AddInstance(inst)
    self:SendToList(inst, self.updaters)
end

function StateGraphWrangler:Update(current_tick)
    local waiters = self.tickwaiters[current_tick]
    if waiters then
        for k,v in pairs(waiters) do
            self.updaters[k] = true
            self.instances[k] = self.updaters
        end
        self.tickwaiters[current_tick] = nil
    end

    local updaters = self.updaters
    self.updaters = {}
    
    TheSim:ProfilerPush("updaters")
    for k,v in pairs(updaters) do
        if k.inst:IsValid() then
            TheSim:ProfilerPush(tostring(k.inst.prefab) .. " > " .. tostring(k.currentstate.name) )
            local sleep_amount = k:Update()
            TheSim:ProfilerPop()
            if sleep_amount then
                if sleep_amount > 0 then
                  self:Sleep(k, sleep_amount)
                else
                    self.updaters[k] = true
                    self.instances[k] = self.updaters
                end
            else
                self:Idle(k)
            end
        end
    end
    TheSim:ProfilerPop()
    
    local evs = self.haveEvents
    self.haveEvents = {}
    
    TheSim:ProfilerPush("events")
    for k,v in pairs(evs) do
        k:HandleEvents()
    end
    TheSim:ProfilerPop()
end

ActionHandler = Class(
    function(self, action, state, condition)
        
        self.action = action
        
        if type(state) == "string" then
            self.deststate = function(inst) return state end
        else
            self.deststate = state
        end
        
        self.condition = condition
    end)

EventHandler = Class(
    function(self, name, fn)
        local info = debug.getinfo(3, "Sl")
        self.defline = string.format("%s:%d", info.short_src, info.currentline)
        assert (type(name) == "string")
        assert (type(fn) == "function")
        self.name = string.lower(name)
        self.fn = fn
    end)
    
TimeEvent = Class(
    function(self, time, fn)
        local info = debug.getinfo(3, "Sl")
        self.defline = string.format("%s:%d", info.short_src, info.currentline)
        assert (type(time) == "number")
        assert (type(fn) == "function")
        self.time = time
        self.fn = fn
    end)    
    
State = Class(
    function(self, args)
        local info = debug.getinfo(3, "Sl")
        self.defline = string.format("%s:%d", info.short_src, info.currentline)
        
        assert(args.name, "State needs name")
        self.name = args.name
        self.onenter = args.onenter
        self.onexit = args.onexit
        self.onupdate = args.onupdate
        self.ontimeout = args.ontimeout
        
        self.tags = {}
        if args.tags then
            for k, v in ipairs(args.tags) do
                self.tags[v] = true
            end
        end
        
        self.events = {}
        if args.events ~= nil then
            for k,v in pairs(args.events) do
                assert(v:is_a(EventHandler), "non-EventHandler in event list")
                self.events[v.name] = v
            end
        end
        
        self.timeline = {}
        if args.timeline ~= nil then
            for k,v in ipairs(args.timeline) do
                assert(v:is_a(TimeEvent), "non-TimeEvent in timeline")
                table.insert(self.timeline, v)
            end
        end
        
        local function pred(a,b)
            return a.time < b.time
        end
        table.sort(self.timeline, pred)
        
    end
)

function State:HandleEvent(sg, eventname, data)
    if not data or not data.state or data.state == self.name then
        local handler = self.events[eventname]
        if handler ~= nil then
            return handler.fn(sg.inst, data)
        end
    end
    return false
end
   



StateGraph = Class( function(self, name, states, events, defaultstate, actionhandlers)
    assert(name and type(name) == "string", "You must specify a name for this stategraph")
    local info = debug.getinfo(3, "Sl")
    self.defline = string.format("%s:%d", info.short_src, info.currentline)
    self.name = name
    self.defaultstate = defaultstate
    
    --reindex the tables
    self.actionhandlers = {}
    if actionhandlers then
        for k,v in pairs(actionhandlers) do
            assert( v:is_a(ActionHandler),"Non-action handler added in actionhandler table!")
            self.actionhandlers[v.action] = v
        end
    end
	for k,modhandlers in pairs(ModManager:GetPostInitData("StategraphActionHandler", self.name)) do
		for i,v in ipairs(modhandlers) do
			assert( v:is_a(ActionHandler),"Non-action handler added in mod actionhandler table!")
			self.actionhandlers[v.action] = v
		end
	end

    self.events = {}
    for k,v in pairs(events) do
        assert( v:is_a(EventHandler),"Non-event added in events table!")
        self.events[v.name] = v
    end
	for k,modhandlers in pairs(ModManager:GetPostInitData("StategraphEvent", self.name)) do
		for i,v in ipairs(modhandlers) do
			assert( v:is_a(EventHandler),"Non-event added in mod events table!")
			self.events[v.name] = v
		end
    end

    self.states = {}
    for k,v in pairs(states) do
        assert( v:is_a(State),"Non-state added in state table!")
        self.states[v.name] = v
    end
	for k,modhandlers in pairs(ModManager:GetPostInitData("StategraphState", self.name)) do
		for i,v in ipairs(modhandlers) do
			assert( v:is_a(State),"Non-state added in mod state table!")
			self.states[v.name] = v
		end
    end

	-- apply mods
	local modfns = ModManager:GetPostInitFns("StategraphPostInit", self.name)
	for i,modfn in ipairs(modfns) do
		modfn(self)
	end
end)

function StateGraph:__tostring()
    return "Stategraph : "..self.name--.. " (currentstate="..self.currentstate.name..":"..self.timeinstate..")"
end
    
    
StateGraphInstance = Class( function (self, stategraph, inst)
    self.sg = stategraph
    self.currentstate = nil
    self.timeinstate = 0
    self.lastupdatetime = 0
    self.timelineindex = nil
    self.prevstate = nil
    self.bufferedevents={}
    self.inst = inst
    self.statemem = {}
    self.mem = {}
    self.statestarttime = 0
end)

function StateGraphInstance:__tostring()
    local str =  string.format([[sg="%s", state="%s", time=%2.2f]], self.sg.name, self.currentstate and self.currentstate.name or "<nil>", GetTime() - self.statestarttime)
    str = str..[[, tags = "]]
	if self.tags ~= nil then
		for k,v in pairs(self.tags) do
			str = str..tostring(k)..","
		end
	end
    str = str..[["]]
    return str
end
    
function StateGraphInstance:GetTimeInState()
    return GetTime() - self.statestarttime
end

function StateGraphInstance:PlayRandomAnim(anims, loop)
    local idx = math.floor(math.random() * #anims)
    self.inst.AnimState:PlayAnimation(anims[idx+1], loop)
end

function StateGraphInstance:PushEvent(event, data)
    if data then
        data.state = self.currentstate.name
    else
        data = {state = self.currentstate.name}
    end
    table.insert(self.bufferedevents, {name=event, data=data})
end

function StateGraphInstance:IsListeningForEvent(event)
    return self.currentstate.events[event] ~= nil or self.sg.events[event] ~= nil
end


function StateGraphInstance:StartAction(bufferedaction)
    if self.sg.actionhandlers then
        local handler = self.sg.actionhandlers[bufferedaction.action]
        if handler then
            if not handler.condition or handler.condition(self.inst) then
                if handler.deststate then
                    local state = handler.deststate(self.inst, bufferedaction)
                    if state then
                        self:GoToState(state)
                    else
                        return
                    end
                else
                    self.inst:PerformBufferedAction()
                end
                    
                return true
            end
        end
    end
end

function StateGraphInstance:HandleEvents()
    assert(self.currentstate ~= nil, "we are not in a state!")
    
    if self.inst:IsValid() then
		for k, event in ipairs(self.bufferedevents) do
			if not self.currentstate:HandleEvent(self, event.name, event.data) then
				local handler = self.sg.events[event.name]
				if handler ~= nil then
					handler.fn(self.inst, event.data)
				end
			end
		end
	end
	
    self.bufferedevents = {}
end

function StateGraphInstance:InNewState()
	return self.laststate ~= self.currentstate
end

function StateGraphInstance:GoToState(statename, params)
    local state = self.sg.states[statename]
    
    if not state then
		print (self.inst, "TRIED TO GO TO INVALID STATE", statename)
		return 
    end
    --assert(state ~= nil, "State not found: " ..tostring(self.sg.name).."."..tostring(statename) )
    

    self.prevstate = self.currentstate
    if self.currentstate ~= nil and self.currentstate.onexit ~= nil then 
        self.currentstate.onexit(self.inst, statename)
    end

    -- Record stats
    if METRICS_ENABLED and self.inst == GetPlayer() and self.currentstate and not IsAwayFromKeyBoard() then
        local dt = GetTime() - self.statestarttime
        self.currentstate.totaltime = self.currentstate.totaltime and (self.currentstate.totaltime + dt) or dt  -- works even if currentstate.time is nil
        -- dprint(self.currentstate.name," time in state= ", self.currentstate.totaltime)
    end

    self.statemem = {}
    self.tags = {}
    if state.tags then
        for i,k in pairs(state.tags) do
            self.tags[i] = true
        end
    end
    self.timeout = nil
    self.laststate = self.currentstate
    self.currentstate = state
    self.timeinstate = 0

    if self.currentstate.timeline ~= nil then
        self.timelineindex = 1
    else
        self.timelineindex = nil
    end
    
    if self.currentstate.onenter ~= nil then
        self.currentstate.onenter(self.inst, params)
    end
    
    self.inst:PushEvent("newstate", {statename = statename})
    
    
    self.lastupdatetime = GetTime()
    self.statestarttime = self.lastupdatetime    
    SGManager:OnEnterNewState(self)

end

function StateGraphInstance:AddStateTag(tag)
    self.tags[tag] = true
end

function StateGraphInstance:RemoveStateTag(tag)
    self.tags[tag] = nil
end

function StateGraphInstance:HasStateTag(tag)
    return self.tags and (self.tags[tag] == true)
end

function StateGraphInstance:SetTimeout(time)
    SGManager:OnSetTimeout(self)
    self.timeout = time
end

function StateGraphInstance:UpdateState(dt)
    if not self.currentstate then 
        return
    end

    self.timeinstate = self.timeinstate + dt
    local startstate = self.currentstate
    
    
    if self.timeout then
        self.timeout = self.timeout - dt
        if self.timeout <= (1/30) then
            self.timeout = nil
            if self.currentstate.ontimeout then
                self.currentstate.ontimeout(self.inst)
                if startstate ~= self.currentstate then
                    return
                end
            end
        end
    end
    
    while self.timelineindex and self.currentstate.timeline[self.timelineindex] and self.currentstate.timeline[self.timelineindex].time <= self.timeinstate do

		local idx = self.timelineindex
        self.timelineindex = self.timelineindex + 1
        if self.timelineindex > #self.currentstate.timeline then
            self.timelineindex = nil
        end
        
        local old_time = self.timeinstate
        local extra_time = self.timeinstate - self.currentstate.timeline[idx].time
        self.currentstate.timeline[idx].fn(self.inst)
        
        
        if startstate ~= self.currentstate or old_time > self.timeinstate then
            self:Update(extra_time)
            return 0
        end
    end
    
    if self.currentstate.onupdate ~= nil then
        self.currentstate.onupdate(self.inst, dt)
    end
end

function StateGraphInstance:Start()
    if self.OnStart then
        self:OnStart()
    end
    self.stopped = false
    SGManager:AddInstance(self)
end

function StateGraphInstance:Stop()
    self:HandleEvents()
    if self.OnStop then
        self:OnStop()
    end
    self.stopped = true
    SGManager:RemoveInstance(self)
end

function StateGraphInstance:Update()
    local dt = 0
    if self.lastupdatetime then
        dt = GetTime() - self.lastupdatetime --+ GetTickTime()
    end
    self.lastupdatetime = GetTime()
	
	
    self:UpdateState(dt)
	
   
    local time_to_sleep = nil
    if self.timelineindex and self.currentstate.timeline and self.currentstate.timeline[self.timelineindex] then
        time_to_sleep = self.currentstate.timeline[self.timelineindex].time - self.timeinstate
    end
        
    
    if self.timeout and (not time_to_sleep or time_to_sleep > self.timeout) then
        time_to_sleep = self.timeout
    end
        
    if self.currentstate.onupdate then
        return 0
    elseif time_to_sleep then
        return time_to_sleep
    else
        return nil
    end
end

    

File: Prototype/stategraph.py
# coding=utf-8
# stategraph.py
# created on 2020/8/18
# author @zoloypzuo
# usage: 状态机
"""
构造状态机：Prefab构造中调用EntityScript.set_stategraph来实例化StateGraphInstance

"""
from common.zlogger import logger_tail, logger
from common.ztable import Table
from common.zutils import is_iterable_T
from game_timer import now


class StateGraphComponent(object):
    """
    SG移植到Messiah，以组件的形式
    Entity继承StateGraphComponent以获得SG
    """

    def __init__(self):
        super(StateGraphComponent, self).__init__()
        self.sg = None

    @logger
    def set_stategraph(self, name, sg_base='stategraphs.'):
        """

        :param sg_base:
        :param self: EntityScript
        :param name: 状态机名字，位于stategraphs/下的文件名
        :return:
        """
        if self.sg:
            SGManager.remove_instance(self.sg)
        sg = load_stategraph(sg_base + name, name)
        assert sg, "load stategraph failed"
        self.sg = StateGraphInstance(sg, self)
        SGManager.add_instance(self.sg)
        self.sg.go_to_state(sg.default_state)
        self.sg.start()  # TODO
        return self.sg

    @logger
    def clear_stategraph(self):
        if self.sg:
            self.sg.stop()  # TODO
            SGManager.remove_instance(self.sg)
            self.sg = None


class StateGraphInstance(object):
    """
    运行时状态机实例
    sg和inst由ctor参数初始化，其他是运行时需要并创建的数据结构
    Attributes:
        sg: StateGraph
        inst: EntityScript
        is_stopped: stop调用后为True
        on_start: 事件, 0 ref
        on_stop: 事件, 0 ref

        current_state: /
        last_state: /

        state_start_time: 当前状态的开始时间，从切状态开始

        buffered_events: 事件队列
        tags: 当前状态的tags，切状态时由State.tag的拷贝初始化，之后可以动态添加移除tag
        events: /

    """

    def __init__(self, sg, inst, on_start=None, on_stop=None):
        super(StateGraphInstance, self).__init__()
        self.sg = sg
        self.inst = inst

        self.current_state = None
        self.last_state = None
        self.state_start_time = 0
        self.buffered_events = []
        self.tags = set()
        self.events = {}

        self.on_start = on_start
        self.on_stop = on_stop

        self.is_stopped = False
        self.timeline_index = None

    # ---------------------------------------------------
    # property
    # ---------------------------------------------------
    @property
    def time_in_state(self):
        """
        当前状态开始到现在的时间差
        :return:
        """
        return now() - self.state_start_time

    # ---------------------------------------------------
    # goto-state
    # ---------------------------------------------------
    @logger
    def go_to_state(self, state_name, *params):
        """
        /
        :param state_name:
        :param params: Table
        :return:
        """
        assert state_name in self.sg.states, "TRIED TO GO TO INVALID STATE %s" % state_name
        from_state = self.current_state
        from_state and from_state.on_exit and from_state.on_exit(self.inst, state_name)
        to_state = self.sg.states[state_name]
        self.tags = set(to_state.tags) if to_state.tags else set()  # 拷贝新状态的tags
        self.last_state, self.current_state = self.current_state, to_state
        to_state.on_enter and to_state.on_enter(self.inst, params)
        self.inst.push_event('new_state', Table(state_name=state_name))
        self.state_start_time = now()
        self.timeline_index = 0 if to_state.timeline else None
        SGManager.on_enter_new_state(self)

    # ---------------------------------------------------
    # event
    # ---------------------------------------------------
    @logger_tail
    def is_listening_for_event(self, event):
        return event in self.current_state.events or event in self.sg.events

    @logger
    def push_event(self, event, data=None):
        """
        EntityScript.push_event
        :param event:
        :param data:
        :return:
        """
        data = data or Table()
        data.state = self.current_state.name
        self.buffered_events.append(Table(name=event, data=data))

    def handle_events(self):
        """
        处理事件
        StateGraphInstance.Stop时会再处理一次事件
        :return:
        """
        assert self.current_state, "we are not in a state!"
        if not self.inst.is_valid():
            return
        buffered_event = self.buffered_events
        for event in buffered_event:
            res = self.current_state.handle_event(self, event.name, event.data)
            if not res:
                handler = self.sg.events.get(event.name, None)
                handler and handler.fn(self.inst, event.data)
        self.clear_buffered_events()

    def clear_buffered_events(self):
        self.buffered_events = []

    # ---------------------------------------------------
    # tag
    # ---------------------------------------------------
    @logger
    def add_state_tag(self, tag):
        self.tags.add(tag)

    @logger
    def remove_stat_tag(self, tag):
        self.tags.remove(tag)

    # @logger_tail
    def has_state_tag(self, tag):
        return tag in self.tags

    # ---------------------------------------------------
    # start-update-stop
    # ---------------------------------------------------
    @logger
    def start(self):
        """

        :return:
        """
        self.on_start and self.on_start()
        self.is_stopped = False
        SGManager.add_instance(self)

    def _update_state(self, dt):
        """

        :param dt:
        :return:
        """
        if not self.current_state:
            return

        while self.timeline_index:
            idx = self.timeline_index
            timeline = self.current_state.timeline
            self.timeline_index = idx + 1 if idx < len(timeline) else None
            if self.timeline_index and timeline[idx].time <= self.time_in_state:
                time_event = timeline[idx]

        on_update = self.current_state.on_update
        on_update and on_update(self.inst, dt)

    def update(self, dt):
        """

        :return:
        """
        self._update_state(dt)

    @logger
    def stop(self):
        """

        :return:
        """
        self.handle_events()
        self.on_stop and self.on_stop()
        self.is_stopped = False
        SGManager.remove_instance(self)

    def __str__(self):
        return "SGI(%s)" % ', '.join(map(repr, [
            # id(self),
            self.sg.name,
            self.current_state.name if self.current_state else 'un_init',
            float('{:.2f}'.format(self.time_in_state)),
            list(self.tags)]))

    __repr__ = __str__


class StateGraphWrangler(object):
    """
    管理所有StateGraphInstance
    SGInstance要先注册到SGManager，才能触发PushEvent
    """

    def __init__(self):
        self.instances = {}  # sg-inst -> 当前所在表
        self.updaters = set()
        self.tick_waiters = set()
        self.have_events = set()

    def _send_to_list(self, sg_inst, _list):
        """
        将inst从旧表中移, 加入list中
        :param sg_inst:
        :param _list:
        :return:
        """
        old_list = self.instances.get(sg_inst, None)
        old_list and old_list.discard(sg_inst)
        self.instances[sg_inst] = _list
        _list.add(sg_inst)

    def has_instance(self, sg_inst):
        return sg_inst in self.instances

    @logger_tail
    def on_push_event(self, sg_inst):
        if sg_inst in self.instances:
            self.have_events.add(sg_inst)
            return True
        return False

    def update(self, current_tick):
        for updater in self.updaters:
            updater.update(current_tick)

        events = self.have_events
        self.have_events = set()
        for e in events:
            e.handle_events()

    # ---------------------------------------------------
    # 增删inst
    # ---------------------------------------------------
    @logger
    def add_instance(self, sg_inst):
        self._send_to_list(sg_inst, self.updaters)

    @logger
    def remove_instance(self, sg_inst):
        old_list = self.instances.get(sg_inst, None)
        if old_list:
            old_list.discard(sg_inst)
            del self.instances[sg_inst]
        self.updaters.discard(sg_inst)
        self.have_events.discard(sg_inst)

    # ---------------------------------------------------
    # on
    # ---------------------------------------------------
    # @logger
    def on_enter_new_state(self, sg_inst):
        pass

    @logger
    def on_remove_entity(self, inst):
        self.remove_instance(inst.sg)

    def __str__(self):
        return "<SGMgr>"

    __repr__ = __str__


SGManager = StateGraphWrangler()


class State(object):
    """
    Attributes:
        name: /；必须有，"State needs name"
        on_enter: on_enter(EntityScript, params)
        on_exit: on_exit(EntityScript, to_state_name)
        on_update: on_update(EntityScript, dt)
        tags: set<str>，初始化时拷贝一份集合
        events: 事件表，EventHandler.name -> EventHandler
        timeline: 延时事件表，list<TimeEvent>，初始化时按TimeEvent.time排序
    """

    def __init__(self, name,
                 on_enter=None, on_update=None, on_exit=None,
                 tags=None, events=None, timeline=None):
        """
        构造一个状态，名字是必须的，作为标识符，其他都是可选的，默认为None
        # [x] timeline
        :param on_enter: on_enter(inst, params)
        :param on_exit: on_exit(inst, to_state_name)
        :param on_update: on_update(inst, dt)
        """
        super(State, self).__init__()
        assert name, "state needs name"
        self.name = name
        self.on_enter = on_enter
        self.on_update = on_update
        self.on_exit = on_exit

        if tags:
            assert is_iterable_T(tags, str)
        self.tags = set(tags) if tags else set()

        if events:
            assert is_iterable_T(events, EventHandler)
        self.events = {e.name: e for e in events} if events else {}

        if timeline:
            assert is_iterable_T(timeline, TimeEvent)
        self.timeline = sorted(list(timeline), key=lambda o: o.time) if timeline else []

    @logger_tail
    def handle_event(self, sg, event_name, data):
        """
        处理事件
        如果事件状态不是当前状态，则返回处理失败
        否则使用事件表处理该事件（仍然可能处理失败）
        :param sg: StateGraphInstance
        :param event_name:
        :param data: Table
        :return: bool，是否成功处理，是则结束，不是则传递事件给SG层事件表处理
        """
        if data and data.state and data.state != self.name:
            return False
        handler = self.events.get(event_name, None)
        if handler:
            return handler.fn(sg.inst, data)
        return False

    def __str__(self):
        return self.name

    __repr__ = __str__


class EventHandler(object):
    """
    Attributes:
        name: 事件名，事件标识符
        fn: fn(sg.inst, data) -> bool
            sg.inst is EntityScript
            返回是否成功处理
            成功处理，事件将停止传递给SG层事件表处理
            如果不，事件将传递给SG层事件表处理
    """

    def __init__(self, name, fn):
        self.name = name
        self.fn = fn


class TimeEvent(object):
    """
    timeline上延时触发函数
    状态拥有timeline，timeline是TimeEvent列表
    当前状态中才能触发TimeEvent，过了就没了
    Attributes:
        time: 延迟，比如5 * FRAMES，代表延迟5帧
        fn: fn(inst)
    """

    def __init__(self, time, fn):
        self.time = time
        self.fn = fn


class StateGraph(object):
    """
    静态的状态机
    Attributes:
        name: 必须有，"You must specify a name for this stategraph"
        states: 状态表，State.name -> State
        events: SG层事件表，处理State事件表未能成功处理的事件
        default_state: /
    """

    def __init__(self, name, events, states, default_state):
        super(StateGraph, self).__init__()
        assert name, "You must specify a name for this stategraph"
        self.name = name
        self.default_state = default_state
        assert is_iterable_T(events, EventHandler)
        self.events = {e.name: e for e in events} if events else {}
        assert is_iterable_T(states, State)
        self.states = {s.name: s for s in states} if states else {}

    def __str__(self):
        return "State(%s)" % repr(self.name)

    __repr__ = __str__


# @memo
def load_stategraph(path, name):
    """
    加载sg模块

    加载配置文件是程序错误，总应该正确，所以不做异常处理
    """
    import importlib
    sg_mod = importlib.import_module(path)
    return getattr(sg_mod, name)


def set_stategraph(inst, name, sg_base='stategraphs.'):
    """
    :param sg_base:
    :param inst: EntityScript
    :param name: 状态机名字，位于stategraphs/下的文件名
    :return:
    """

    if inst.sg:
        SGManager.remove_instance(inst.sg)
    sg = load_stategraph(sg_base + name, name)
    assert sg, "load stategraph failed"
    inst.sg = StateGraphInstance(sg, inst)
    SGManager.add_instance(inst.sg)
    inst.sg.go_to_state(sg.default_state)
    return inst.sg


def clear_stategraph(inst):
    if inst.sg:
        SGManager.remove_instance(inst.sg)
        inst.sg = None
